File: examples/06-saga-spring/python/continue_as_new.py
from temporalio import workflow

# `has_next_subscription_event` / `handle_next_subscription_event` stand in for
# signal-driven state on the workflow (a Query-able buffer in a real app).


@workflow.defn
class SubscriptionWorkflow:
    # An indefinitely-running, event-driven workflow. After a bounded number of
    # events we call continue_as_new to start a fresh run with a clean history,
    # carrying forward only the state the next run needs.
    @workflow.run
    async def run(self, subscription_id: str, event_count: int) -> None:
        while True:
            await workflow.wait_condition(lambda: self.has_next_subscription_event())
            self.handle_next_subscription_event()
            event_count += 1

            if event_count >= 1000:
                # Does not return — replaces this run with a new one (event_count reset).
                workflow.continue_as_new(args=[subscription_id, 0])

File: examples/06-saga-spring/go/continue_as_new.go
package saga

import (
	"go.temporal.io/sdk/workflow"
)

// SubscriptionWorkflow is an indefinitely-running, event-driven workflow. After a
// bounded number of events it returns a ContinueAsNew error to start a fresh run
// with a clean history, carrying forward only the state the next run needs.
//
// hasNextEvent / handleNextEvent stand in for signal-driven state (a real app
// would read a signal channel here).
func SubscriptionWorkflow(ctx workflow.Context, subscriptionID string, eventCount int) error {
	for {
		_ = workflow.Await(ctx, func() bool { return hasNextEvent(ctx) })
		handleNextEvent(ctx)
		eventCount++

		if eventCount >= 1000 {
			// Replaces this run with a new one (eventCount reset to 0).
			return workflow.NewContinueAsNewError(ctx, SubscriptionWorkflow, subscriptionID, 0)
		}
	}
}
